File: frontend-reactjs/src/features/recipes/utils/validate.js
import { toast } from "react-toastify"
import { validateString, validateStringWithMessage } from "../../../utils/validate/validateString"
import { MAX_DESCRIPTION_LENGTH, MAX_INGREDIENT_NAME_LENGTH, MAX_INGREDIENT_QUANTITY_LENGTH, MAX_INGREDIENTS, MAX_MEDIAS, MAX_STEP_DESCRIPTION_LENGTH, MAX_TAGS, MAX_TITLE_LENGTH, MIN_DESCRIPTION_LENGTH, MIN_INGREDIENT_NAME_LENGTH, MIN_INGREDIENT_QUANTITY_LENGTH, MIN_INGREDIENTS, MIN_STEP_DESCRIPTION_LENGTH, MIN_STEPS, MIN_TITLE_LENGTH } from "../constants/recipeRule"
import { TITLE_VALIDATION_REGEX } from "../constants/validateRegex"

export function validateSubmitRecipe(title, description, image, medias, tags, ingredients, steps) {
    // if (!validateString(title, MIN_TITLE_LENGTH, MAX_TITLE_LENGTH, TITLE_VALIDATION_REGEX)) {
    //     toast.warning("VALIDATE");
    //     return
    // }
    const titleMessage = validateStringWithMessage(title, "Tiêu đề", MIN_TITLE_LENGTH, MAX_TITLE_LENGTH, TITLE_VALIDATION_REGEX, "Tiêu đề chỉ cho phép chữ, số, khoảng trắng và các ký tự: - _ . , ! ? ( )");
    if (titleMessage) {
        toast.warning(titleMessage);
        return false;
    }

    const descriptionMessage = validateStringWithMessage(description, `Mô tả món ăn`, MIN_DESCRIPTION_LENGTH, MAX_DESCRIPTION_LENGTH);
    if (descriptionMessage) {
        toast.warning(descriptionMessage);
        return false; // Thoát toàn bộ function
    }

    if (!image) {
        toast.warning("Chưa có ảnh chính minh họa!");
        return false;
    }
    if (medias.length > MAX_MEDIAS) {
        toast.warning(`Không được quá ${MAX_MEDIAS} ảnh hoặc video`);
        return false;
    }

    if (steps.length < MIN_STEPS) {
        toast.warning(`Số bước làm phải hơn ${MIN_STEPS} bước`);
        return false;
    }

    if (steps.length > MAX_MEDIAS) {
        toast.warning(`Số bước làm không được quá ${MAX_MEDIAS} bước`);
        return false;
    }

    for (let [index, step] of steps.entries()) {
        const stepMessage = validateStringWithMessage(step.description, `Mô tả bước ${index + 1}`, MIN_STEP_DESCRIPTION_LENGTH, MAX_STEP_DESCRIPTION_LENGTH);
        if (stepMessage) {
            toast.warning(stepMessage);
            return false; // Thoát toàn bộ function
        }
    }

    if (tags.length > MAX_TAGS) {
        toast.warning(`Số tag không được quá ${MAX_TAGS} tag`);
        return false;
    }

    if (ingredients.length < MIN_INGREDIENTS) {
        toast.warning(`Số nguyên liệu phải hơn ${MIN_INGREDIENTS} nguyên liệu`);
        return false;
    }

    if (ingredients.length > MAX_INGREDIENTS) {
        toast.warning(`Số nguyên liệu không được quá ${MAX_INGREDIENTS} nguyên liệu`);
        return false;
    }

    for (let [index, ingredient] of ingredients.entries()) {
        const nameMessage = validateStringWithMessage(ingredient.name, `Tên nguyên liệu thứ ${index + 1}`, MIN_INGREDIENT_NAME_LENGTH, MAX_INGREDIENT_NAME_LENGTH);
        if (nameMessage) {
            toast.warning(nameMessage);
            return false; // Thoát toàn bộ function
        }
        const quantityMessage = validateStringWithMessage(ingredient.quantity, `Số lượng của nguyên liệu thứ ${index + 1}`, MIN_INGREDIENT_QUANTITY_LENGTH, MAX_INGREDIENT_QUANTITY_LENGTH);
        if (quantityMessage) {
            toast.warning(quantityMessage);
            return false; // Thoát toàn bộ function
        }
    }
    return true;
}

File: BackendDjango/app/utils/validate.py
from rest_framework import serializers

from app.configs.rules import RecipeRule


def validate_recipe_data(data):

    # Title
    if (not data.get("title") or
            len(data["title"].strip()) < RecipeRule.Title.MIN_TITLE_LENGTH or
            len(data["title"].strip()) > RecipeRule.Title.MAX_TITLE_LENGTH):
        raise serializers.ValidationError(f"Tiêu đề món ăn phải số ký tự từ {RecipeRule.Title.MIN_TITLE_LENGTH} đến {RecipeRule.Title.MAX_TITLE_LENGTH}.")

    if not RecipeRule.Title.TITLE_VALIDATION_REGEX.match(data["title"]):
        raise serializers.ValidationError(
            f"Tiêu đề chỉ cho phép chữ, số, khoảng trắng và các ký tự: - _ . , ! ? ( ).")

    # Description
    # if not data.get("description") or len(data["description"].strip()) < 10:
    #     errors.append("Mô tả món ăn phải có ít nhất 10 ký tự.")
    if (not data.get("description") or
            len(data["description"].strip()) < RecipeRule.Description.MIN_DESCRIPTION_LENGTH or
            len(data["description"].strip()) > RecipeRule.Description.MAX_DESCRIPTION_LENGTH):
        raise serializers.ValidationError(f"Mô tả món ăn phải đúng số lượng ký tự.")


    # Image (bắt buộc)
    print("data.get(image): ", data.get("image"))
    if data.get("image") is None:
        raise serializers.ValidationError("Vui lòng tải ảnh món ăn.")

    # Steps
    if not data.get("steps") or len(data["steps"]) < RecipeRule.Step.MIN_STEPS or len(data["steps"]) > RecipeRule.Step.MAX_STEPS:
        raise serializers.ValidationError("Số bước nấu ăn phải hợp lệ ")
    else:
        for i, step in enumerate(data["steps"], start=1):
            if not step.get("description") or len(step["description"].strip()) < RecipeRule.Step.MIN_STEP_DESCRIPTION_LENGTH:
                raise serializers.ValidationError.append(f"Bước {i} cần mô tả đủ ký tự.")

    # Tags
    # if not data. len(data["tags"]) == 0:
    #     errors.append("Phải có ít nhất 1 tag cho món ăn.")

    # Ingredients
    # if not data.get("ingredients") or len(data["ingredients"]) == 0:
    #     errors.append("Phải có ít nhất 1 nguyên liệu.")
    # else:
    #     for i, ing in enumerate(data["ingredients"], start=1):
    #         if not ing.get("name") or not ing.get("quantity"):
    #             errors.append(f"Nguyên liệu {i} thiếu tên hoặc số lượng.")
